File: Systems/Sensors/BaseSensor.py
import logging
import secrets
from datetime import datetime, timedelta

from Utils.Decorators import try_catch, type_check

logger = logging.getLogger(__name__.split(".")[-1])


class BaseSensor:
    """ BaseSensor class """

    @type_check
    def __init__(self, owner: None, name: str, address: str) -> None:
        """ Initialize an instance of the BaseSensor class.

        Arguments:
            owner {SensorSystem} -- SensorSystem object which is the owner of the sensor.
            name {str} -- Name of the sensor.
            address {str} -- (IP) Address of the sensor.
        """

        self._owner = owner
        self.name = name
        self.address = address
        self.token = secrets.token_hex(16)
        self.metadata = {}  # subName / (type / value)

        self._data = {}  # time / (subName / value)
        self._lastReadings = {}  # subName / value

    @type_check
    def load(self, data: dict) -> None:
        """ Loads sensor's data.

        Arguments:
                data {dict} -- Dictionary from which the sensor data will be loaded.
        """

        logger.debug("Load sensor: %s", self.name)

        if "token" in data:
            self.token = data["token"]
            if self.address == self.token:
                self.address = ""
        if "metadata" in data:
            self.metadata = data["metadata"]
        if "data" in data:
            self._data = data["data"]
        if "lastReadings" in data:
            self._lastReadings = data["lastReadings"]

    @type_check
    def save(self, data: dict) -> None:
        """ Saves sensor's data.

        Arguments:
                data {dict} -- Dictionary to which the sensor data will be saved.
        """

        logger.debug("Save sensor: %s", self.name)

        # Note: address is saved by the config and SensorsSystem's sensors dict
        data["token"] = self.token
        data["metadata"] = self.metadata
        data["data"] = self._data
        data["lastReadings"] = self._lastReadings

    @property
    @type_check
    def latestTime(self) -> datetime:
        """ Gets a latest time data added. """

        if len(self._data) == 0:
            return None

        return sorted(self._data.keys())[-1]

    @property
    @type_check
    def latestData(self) -> dict:
        """ Gets a dictionary with the latest sensor's data. """

        if self.latestTime is None:
            return {}

        return self._data[self.latestTime]

    @property
    @type_check
    def subNames(self) -> list:
        """ Gets a list with all sub names. """

        if self.latestTime is None:
            return []

        return list(self._data[self.latestTime].keys())

    @type_check
    def update(self) -> None:
        """ Update current sensor's state. """
        pass

    @type_check
    def readData(self, time: datetime, addBiggerOnly: bool = False) -> list:
        """ Read and store data from the sensor into data collection.

        Arguments:
            time {datetime} -- Time when the data is collected.
            addBiggerOnly {bool} -- Add only data with bigger value. (default: {False})
        """

        data = self._readData()  # read data from sensor
        logger.debug("Sensor '%s' read data: %s", self.name, data)
        if data is not None:
            self.addData(time, data, addBiggerOnly)
        return data

    @type_check
    def addData(self, time: datetime, data: list, addBiggerOnly: bool = False) -> bool:
        """ Add data to the sensors data collection.

        Arguments:
            time {datetime} -- Time when the data is collected.
            data {list} -- Data which will be added.
            addBiggerOnly {bool} -- Add only data with bigger value. (default: {False})

        Returns:
            bool -- True if data was added, otherwise False.
        """

        if time is None:
            time = datetime.now()

        logger.debug("Sensor '%s' add data at %s: %s (%s)", self.name, time, data, addBiggerOnly)

        result = False
        if time not in self._data:
            self._data[time] = {}
        for item in data:
            if "name" in item and "value" in item:
                aggrType = item["aggrType"] if "aggrType" in item else "avg"
                if aggrType == "avg":
                    if addBiggerOnly and item["name"] in self._data[time] and self._data[time][item["name"]] > item["value"]:
                        # if add only bigger and there is bigger previous value
                        logger.debug("Sensor '%s' skip adding '%s' data with value '%s' smaller than previous '%s'",
                                     self.name, item["name"], item["value"], self._data[time][item["name"]])
                    else:
                        self._data[time][item["name"]] = item["value"]
                        result = True
                else:  # sum type - differentiate
                    prevValue = self._lastReadings[item["name"]
                                                   ] if item["name"] in self._lastReadings else 0
                    if prevValue > item["value"]: # if the new value is smaller than previous - it's reset
                        prevValue = 0
                    self._data[time][item["name"]] = item["value"] - prevValue
                    self._lastReadings[item["name"]] = item["value"]
                    # set real value to return in readData
                    item["value"] -= prevValue
                    result = True
                self.metadata[item["name"]] = {
                    key: value for key, value in item.items() if key not in ("name", "value")}
            else:
                logger.warning(
                    "Try to add invalid data item(%s) in sensor(%s)", item, self.name)

        if result:
            self._owner._owner.event(self, "SensorDataAdded", {
                                    item["name"]: item["value"] for item in data})
            self._archiveData()
        return result

    @type_check
    def _readData(self) -> list:
        """ Abstract method to read data from the sensor. """
        return []

    @try_catch("Cannot archive data")
    @type_check
    def _archiveData(self) -> None:
        """ Archive old data. """

        now = datetime.now()
        # delete entries older then 1 year
        times = [time for time in self._data.keys() if time <
                 now.replace(year=now.year-1)]
        for time in times:
            del self._data[time]

        # for older then 24 hour, save only 1 per day
        times = [time for time in self._data.keys() if time < now.replace(
            hour=0, minute=0, second=0, microsecond=0) - timedelta(days=1)]
        groupedDate = {}
        for time in times:
            if time.date() not in groupedDate:
                groupedDate[time.date()] = []
            groupedDate[time.date()].append(time)

        for date, times in groupedDate.items():
            if len(times) == 1:  # already archived
                continue

            items = [self._data[t] for t in times]
            # delete old records
            for t in times:
                del self._data[t]
            # add one new
            newTime = datetime.combine(date, datetime.min.time())
            self._data[newTime] = {}
            for subName in self.subNames:
                values = [item[subName] for item in items if subName in item]
                if len(values) == 0:
                    continue

                aggrType = "avg"
                if subName in self.metadata and "aggrType" in self.metadata[subName]:
                    aggrType = self.metadata[subName]["aggrType"]

                newValue = None
                if aggrType == "avg":
                    if isinstance(values[0], bool):
                        newValue = len([v for v in values if v]) >= float(
                            len(values)) / 2  # if True values are more then False
                    elif isinstance(values[0], int):
                        newValue = int(round(sum(values) / float(len(values))))
                    elif isinstance(values[0], float):
                        newValue = sum(values) / float(len(values))
                elif aggrType == "sum":
                    if isinstance(values[0], bool):
                        newValue = len([v for v in values if v]) > 0
                    else:
                        newValue = sum(values)
                self._data[newTime][subName] = newValue

File: Systems/Devices/Sensors/BaseSensor.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

using MyHome.Models;
using MyHome.Utils;

using Newtonsoft.Json;

using NLog;

namespace MyHome.Systems.Devices.Sensors
{
    public abstract class BaseSensor : Device
    {
        private static readonly Logger logger = LogManager.GetCurrentClassLogger();

        public class SensorValue : Dictionary<string, double> { }; // subName / value

        public enum AggregationType
        {
            AverageByTime,
            Average,
            SumDiff, // differentiate income values and sum them on aggregate
            Sum // sum incoming values on aggregate
        }

        public override DateTime LastOnline => this.Data.Keys.OrderBy(t => t).LastOrDefault();


        [JsonIgnore]
        public Dictionary<DateTime, SensorValue> Data { get; }

        [UiProperty(true, "real name / custom name")]
        public Dictionary<string, string> SubNamesMap { get; } // map sensor subname to custom subname

        [UiProperty(true, "name / aggregation type")]
        public Dictionary<string, AggregationType> AggregationMap { get; } // subname / aggregation type

        [UiProperty(true, "name / expression of x")]
        public Dictionary<string, string> Calibration { get; } // calibration values per subname: newValue = expression(realValue)

        [UiProperty(true, "name / unit")]
        public Dictionary<string, string> Units { get; } // subname / unit name

        [UiProperty(true, "name")]
        public List<string> NotTimeseries { get; } // subnames that need code to generate intermediate values

        [UiProperty(true)]
        public bool Grouped { get; set; } // whether subnames should be grouped in UI 


        [JsonProperty]
        private Dictionary<string, double> LastReadings { get; } // subname / value


        [JsonIgnore]
        public Dictionary<string, Dictionary<string, object>> Metadata => this.GetMetadata();

        [JsonIgnore]
        [UiProperty]
        public Dictionary<string, double> Values => this.GetValues();


        [JsonProperty]
        protected string dataFilePath;
        private DateTime? saveTime;
        [JsonProperty]
        private DateTime lastBackupTime;


        protected BaseSensor()
        {
            this.Data = new Dictionary<DateTime, SensorValue>();
            this.SubNamesMap = new Dictionary<string, string>();
            this.AggregationMap = new Dictionary<string, AggregationType>();
            this.Calibration = new Dictionary<string, string>();
            this.Units = new Dictionary<string, string>();
            this.NotTimeseries = new List<string>();
            this.Grouped = false;

            this.LastReadings = new Dictionary<string, double>();
            this.dataFilePath = null;

            this.saveTime = null;
            this.lastBackupTime = DateTime.Now;
        }

        public override void Setup()
        {
            base.Setup();

            if (File.Exists(this.dataFilePath))
            {
                try
                {
                    try
                    {
                        var json = File.ReadAllText(this.dataFilePath);
                        JsonConvert.PopulateObject(json, this.Data);
                    }
                    catch
                    {
                        // retry with the copy file
                        logger.Warn($"Cannot load sensor data '{this.Name}' ({this.Room.Name}), retry with the copy");
                        var json2 = File.ReadAllText(this.dataFilePath + "1");
                        JsonConvert.PopulateObject(json2, this.Data);
                    }
                    // TODO: remove data older than 365 days, replace save with append, file format without opening and closing json brackets
                }
                catch (Exception e)
                {
                    logger.Error($"Cannot load sensor data '{this.Name}' ({this.Room.Name})");
                    logger.Debug(e);
                    Alert.Create("Cannot load sensor data")
                        .Details($"'{this.Name}' ({this.Room.Name})")
                        .Validity(TimeSpan.FromHours(1))
                        .Send();
                }
            }
        }

        public override void Stop()
        {
            base.Stop();

            // ensure saved on stop
            this.saveTime = DateTime.Now - TimeSpan.FromSeconds(1);
            this.Update();
        }

        public override void Update()
        {
            base.Update();

            if (DateTime.Now > this.saveTime)
            {
                this.saveTime = null;

                logger.Trace($"Save sensor '{this.Name}' ({this.Room.Name}) data");
                var fileName = Path.Join(Config.BinPath, Utils.Utils.GetValidFileName($"data_{this.Room.Name}_{this.Name}.json"));
                this.dataFilePath ??= fileName; // initial set if  DataFilePath is null
                if (this.dataFilePath != fileName) // if there is a change in the name (room/sensor rename)
                {
                    if (File.Exists(fileName))
                        logger.Warn($"Sensor data file already exists: {fileName}");
                    else
                    {
                        if (File.Exists(this.dataFilePath))
                            File.Move(this.dataFilePath, fileName);
                        this.dataFilePath = fileName;
                    }
                }

                // backup data file every day
                if (DateTime.Now - this.lastBackupTime > TimeSpan.FromDays(1))
                {
                    this.lastBackupTime = DateTime.Now;
                    if (File.Exists(this.dataFilePath))
                        File.Copy(this.dataFilePath, this.dataFilePath + ".bak", true);
                }

                lock (this.Data)
                {
                    Utils.Utils.Retry(_ =>
                    {
                        var formatting = MyHome.Instance.Config.SavePrettyJson ? Formatting.Indented : Formatting.None;
                        var json = JsonConvert.SerializeObject(this.Data, formatting);
                        File.WriteAllText(this.dataFilePath, json);
                        File.WriteAllText(this.dataFilePath + "1", json); // save a copy if the first one is broken
                    }, 3, logger, $"save sensor '{this.Name}' ({this.Room.Name}) data");
                }
            }
        }

        public virtual void SaveData()
        {
            // save only once per minute
            if (this.saveTime == null)
                this.saveTime = DateTime.Now + TimeSpan.FromMinutes(1);
        }


        public void AddData(DateTime time, Dictionary<string, object> data)
        {
            lock (this.Data)
            {
                logger.Trace($"Sensor '{this.Name}' ({this.Room.Name}) add data at {time:dd/MM/yyyy HH:mm:ss}: {string.Join('\n', data)}");
                if (data.Count == 0)
                    return;

                if (!this.Data.ContainsKey(time))
                    this.Data.Add(time, new SensorValue());

                foreach (var item in data)
                {
                    var name = item.Key;
                    if (this.SubNamesMap.ContainsKey(name))
                        name = this.SubNamesMap[name];

                    var value = Convert.ToDouble(item.Value);
                    if (this.Calibration.ContainsKey(name)) // mapped name
                    {
                        MyHome.Instance.ExecuteJint(jint => value = (double)jint.SetValue("x", value).Evaluate(this.Calibration[name]).ToObject(),
                            "calculate value: " + this.Calibration[name]);
                    }

                    this.AggregationMap.TryGetValue(name, out var aggrType);
                    if (aggrType != AggregationType.SumDiff)
                        this.Data[time][name] = value;
                    else // sum diff type - differentiate
                    {
                        this.LastReadings.TryGetValue(name, out var prevValue);
                        // TODO: uncomment once the Energy Monitor values are fixed
                        //if (prevValue - 1 > value) // if the new value is smaller (with epsilon - 1) than previous - it's reset
                        //    prevValue = 0;
                        if (prevValue + 1 < value)
                            this.Data[time][name] = Math.Round(value - prevValue, 4); // round to 4 decimals after the point
                        else
                            this.Data[time][name] = 0;
                        this.LastReadings[name] = value;
                    }
                }
                this.Data[time].TrimExcess();
                MyHome.Instance.Events.Fire(this, GlobalEventTypes.SensorDataAdded, this.Data[time]);
                this.ArchiveData();
                this.SaveData();
            }
        }

        public void GenerateTimeseries()
        {
            lock (this.Data)
            {
                // if subsensor is not timeseries, add last value to fill the gaps in time
                var now = DateTime.Now;
                foreach (var subname in this.NotTimeseries.Distinct())
                {
                    if (!this.Data.ContainsKey(now))
                        this.Data[now] = new SensorValue();
                    this.Data[now].Add(subname, this.Values.ContainsKey(subname) ? this.Values[subname] : 0);
                }
                this.ArchiveData();
                this.SaveData();
            }
        }

        public void AggregateData(IEnumerable<IGrouping<DateTime, DateTime>> groupedDates)
        {
            lock (this.Data)
            {
                foreach (var group in groupedDates) // new time / list of times to be grouped
                {
                    if (group.Count() == 1) // already grouped
                        continue;

                    var items = group.ToDictionary(t => t, t => this.Data[t]);
                    // delete old records
                    foreach (var t in group)
                        this.Data.Remove(t);
                    // add one new
                    this.Data[group.Key] = new SensorValue();
                    var subNames = items.Select(i => i.Value.Keys).SelectMany(x => x).Distinct();
                    foreach (var subName in subNames)
                    {
                        var values = items.Where(i => i.Value.ContainsKey(subName)).ToDictionary(i => i.Key, i => i.Value[subName]);
                        if (!values.Any())
                            continue;

                        this.AggregationMap.TryGetValue(subName, out var aggrType);

                        double newValue = 0.0;
                        if (aggrType == AggregationType.Average)
                            newValue = Math.Round(values.Values.Average(), 4);
                        else if (aggrType == AggregationType.AverageByTime)
                            newValue = Math.Round(AverageByTime(values), 4);
                        else // AggregationType.SumDiff and AggregationType.Sum
                            newValue = Math.Round(values.Values.Sum(), 4);
                        this.Data[group.Key][subName] = newValue;
                    }
                    this.Data[group.Key].TrimExcess();
                }
            }
        }

        private void ArchiveData()
        {
            this.RemoveUnwantedSubData();

            var now = DateTime.Now;
            // TODO: don't delete for now to see how file size will be
            // delete entries older then 3 year
            /*var times = this.Data.Keys.Where(t => t < now.AddYears(-3));
            foreach (var time in times)
                this.Data.Remove(time);*/

            // for older then 24 hour, save only 1 per day
            var times = this.Data.Keys.Where(t => t < now.Date.AddDays(-1));
            var groupedDates = times.GroupBy(t => t.Date);
            this.AggregateData(groupedDates);
        }

        private void RemoveUnwantedSubData()
        {
            // remove subdata with subname not defined in SubNamesMap
            var unwantedDataSubNames = this.Values.Keys.Where(subName => !this.SubNamesMap.ContainsValue(subName));
            if (!unwantedDataSubNames.Any())
                return;

            logger.Warn($"The following '{this.Name}' sensor sub-data will be removed: {string.Join(", ", unwantedDataSubNames)}");
            foreach (var date in this.Data.Keys)
            {
                foreach (var subName in unwantedDataSubNames)
                    this.Data[date].Remove(subName);
            }

            foreach (var subName in unwantedDataSubNames)
                this.LastReadings.Remove(subName);
        }

        private Dictionary<string, Dictionary<string, object>> GetMetadata()
        {
            var result = new Dictionary<string, Dictionary<string, object>>();
            foreach (var key in this.Values.Keys)
            {
                this.AggregationMap.TryGetValue(key, out var aggrType);
                var info = new Dictionary<string, object>
                {
                    { "aggrType", aggrType.ToString() },
                    { "LastOnline", this.LastOnline }
                };
                if (this.Units.ContainsKey(key))
                    info.Add("unit", this.Units[key]);
                result.Add(key, info);
            }
            return result;
        }

        private Dictionary<string, double> GetValues()
        {
            var result = this.Data.OrderBy(kvp => kvp.Key)
                .SelectMany(kvp => kvp.Value)
                .GroupBy(x => x.Key)
                .ToDictionary(g => g.Key, g => g.Last().Value);

            // for sum type sensors - set value to be the sum for the day
            var sumType = this.Data.Where(kvp => kvp.Key > DateTime.Now.Date)
                .SelectMany(kvp => kvp.Value)
                .GroupBy(x => x.Key)
                .Where(g => this.AggregationMap.ContainsKey(g.Key) && this.AggregationMap[g.Key] == AggregationType.SumDiff)
                .ToList();
            sumType.ForEach(g => result[g.Key] = g.Select(x => x.Value).Sum());
            return result;
        }

        private static double AverageByTime(Dictionary<DateTime, double> values)
        {
            if (values.Count <= 2)
                return values.Values.Average();

            double result = 0.0;
            var ordered = values.OrderBy(kvp => kvp.Key).ToList();
            for (int i = 0; i < ordered.Count - 1; i++)
            {
                var duration = (ordered[i + 1].Key - ordered[i].Key).TotalSeconds;
                result += ordered[i].Value * duration;
            }
            // add last value multiplied by average duration
            var totalTime = (ordered[^1].Key - ordered[0].Key).TotalSeconds;
            var avgTime = totalTime / (ordered.Count - 1);
            result += ordered[^1].Value * avgTime;

            return result / (totalTime + avgTime);
        }
    }
}